lib/api/src/modules/stories.ts

```diff
@@ -368,10 +375,36 @@ export const init: ModuleFn = ({
       }
     });
 
-    fullAPI.on(STORY_ARGS_UPDATED, (id: StoryId, args: Args) => {
-      const { storiesHash } = store.getState();
-      (storiesHash[id] as Story).args = args;
-      store.setState({ storiesHash });
+    fullAPI.on(STORY_ARGS_UPDATED, function handleStoryArgsUpdated({
+      storyId,
+      args,
+    }: {
+      storyId: StoryId;
+      args: Args;
+    }) {
+      // the event originates from an iframe, event.source is the iframe's location origin + pathname
+      const { source }: { source: string } = this;
+      const [sourceType, sourceLocation] = getSourceType(source);
+
+      switch (sourceType) {
+        case 'local': {
+          const { storiesHash } = store.getState();
+          (storiesHash[storyId] as Story).args = args;
+          store.setState({ storiesHash });
+          break;
+        }
+        case 'external': {
+          const refs = fullAPI.getRefs();
+          const ref = fullAPI.findRef(sourceLocation);
+          (ref.stories[storyId] as Story).args = args;
+          store.setState({ refs: { ...refs, [ref.id]: ref } });
+          break;
```

Comment:
@ndelangen it feels a bit weird to be changing `refs` inside `modules/stories`. But then moving this logic to refs would be spreading the exact same logic (updating args) into two places 🤔 

I think we need to make refs/stories the same thing already!

Reply:
You want to do that before the RC?

I absolutely agree that's where we want to be at some point. I was just being careful of when we do that.

Reply:
No I don't really want to do it now. Just wondering what to do about the above code.

Reply:
You could consider using fullAPI.setRef
https://github.com/storybookjs/storybook/blob/9a186bae80359c244a45ed94df47dcb566a51707/lib/api/src/modules/refs.ts#L26-L27
?

Reply:
@ndelangen the implementation of that function expects `stories` to be passed in directly from `SET_STORIES`. Can I refactor it?

Reply:
@ndelangen https://github.com/storybookjs/storybook/pull/11401/commits/38e66ad7cb59885eb1b6da7fd47aae7c3376a314 -- is that good for you?